opm/grid/cpgrid/LgrOutputHelpers.cpp

```diff
@@ -0,0 +1,130 @@
+    sorted_levelIdxToLevelCartIdx.reserve(lgr_cells);
+
+    for (const auto& element : Dune::elements(grid.levelGridView(level))) {
+        sorted_levelIdxToLevelCartIdx.push_back(std::make_pair(element.index(),
+                                                               levelCartMapp.cartesianIndex(element.index(),level)));
+    }
+
+    std::sort(sorted_levelIdxToLevelCartIdx.begin(), sorted_levelIdxToLevelCartIdx.end(),
+              [](std::pair<int,int>& a, std::pair<int,int>& b) {
+                  return a.second < b.second;
+              });
+
+    std::vector<int> toOutput; // Consecutive numbers, from 0 to total elemts in LGR1 -1
+    toOutput.reserve(lgr_cells);
+
+    // Redefinition of level Cartesian indices (output-style)
+    for (const auto& [sorted_elemIdx, sorted_cartIdx] : sorted_levelIdxToLevelCartIdx) {
+        toOutput.push_back(sorted_elemIdx);
+    }
+    return toOutput;
+}
+
+void extractSolutionLevelGrids(const Dune::CpGrid& grid,
+                               const Opm::data::Solution& leafSolution,
+                               std::vector<Opm::data::Solution>& levelSolutions)
+
+{
+    int maxLevel = grid.maxLevel();
+
+    // To restrict/create the level cell data, based on the leaf cells and the hierarchy
+    levelSolutions.resize(maxLevel+1);
+
+    for (const auto& [name, leafCellData] : leafSolution)
+    {
+        const auto& measure =  leafCellData.dim;  // Opm::UnitSystem::measure;
+        const auto& target = leafCellData.target; // Opm::data::TargetType>;
+
+        leafCellData.visit([&](const auto& leafVector) {
+            using T = std::decay_t<decltype(leafVector)>;
+
+            if constexpr (std::is_same_v<T, std::monostate>) {
+                // do nothing
+            }
+            else {
+
+                if (!leafVector.empty()) {
+                    std::vector<T> levelVectors{};
+                    levelVectors.resize(maxLevel+1);
+
+                    const auto rubbish = -1;
+                    for (int level = 0; level <= maxLevel; ++level) {
+                        levelVectors[level].resize(grid.levelGridView(level).size(0), rubbish);
+                    }
+
+                    // For level cells that appear in the leaf, extract the data value from leafVector
+                    // and assign it the the equivalent level cell.
+                    // Notice that cells that vanished (parent cells) get the rubbish value.
+                    // Store in the order expected by outout files (increasing level Cartesian indices)
+                    const Opm::LevelCartesianIndexMapper<Dune::CpGrid> levelCartMapp(grid);
+                    for (const auto& element : Dune::elements(grid.leafGridView())) {
+                        int levelCartIdx = levelCartMapp.cartesianIndex(element.getLevelElem().index(), element.level());
+                        levelVectors[element.level()][levelCartIdx] = leafVector[element.index()];
```

Comment:
Here you are assuming that there are not inactive cells.

We probably should not do that. Don't we have the possibility to create a map from local index in the lgr to expected index of output (reorderForOutput)?